File: lib/utils/title_year_parser.dart
/// Parses "raw" IPTV titles into a clean (title, year) pair usable for
/// external-metadata lookups such as TMDB.
///
/// The IPTV provider names come in many shapes:
///   "FR| Fire Country - 2025"
///   "FR| Fire Country (2025)"
///   "SD| Cold Wallet FHD - 2023"
///   "FHD Fire Country"
///   "=FR= FRANCE =FR="    (noise → no clean title)
///
/// The parser is defensive: unrecognised titles are returned as-is with a
/// null year.
class TitleYearParser {
  TitleYearParser._();

  /// Strips IPTV prefixes, quality suffixes and trailing year, returning a
  /// [TitleYear] record. The original string is always preserved in
  /// [TitleYear.original].
  static TitleYear parse(String raw) {
    if (raw.trim().isEmpty) {
      return const TitleYear(title: '', year: null, original: '');
    }

    String s = raw.trim();
    // Drop leading pipe-wrapped tag blocks: "|FR|", "|FR-4K DV|", "|IT-4K|",
    // and stacked ones like "|VO|STFR|". Some panels wrap the country/quality
    // tag in pipes (leading pipe first), e.g. "|FR| Extravagances (1995)".
    // Without this the "FR" leaks into the title and breaks TMDB matching.
    // Looped so multiple stacked blocks are all removed.
    final leadingPipeTag = RegExp(r'^\s*\|[^|]*\|\s*');
    while (leadingPipeTag.hasMatch(s)) {
      s = s.replaceFirst(leadingPipeTag, '');
    }
    // Drop a remaining country / quality prefix without a leading pipe:
    // "FR|", "SD|", "STFR|", "VO|", "BE|", "UK|", "US|" …
    s = s.replaceFirst(
      RegExp(r'^[A-Z]{2,4}[-\s]?\|\s*', caseSensitive: false),
      '',
    );
    // Cut trailing junk after a "||" separator, e.g.
    // "La casa de papel (2017) || MULTI".
    final dbl = s.indexOf('||');
    if (dbl >= 0) s = s.substring(0, dbl).trim();

    // Extract the year. Prefer a parenthesised/bracketed 4-digit year anywhere
    // ("(2017)", "[2020]") — robust against trailing tags like "(VFQ)" that
    // follow the year. Fall back to a bare trailing year (" - 2025", " 2025").
    int? year;
    final paren = RegExp(r'[\(\[](19|20)\d{2}[\)\]]').firstMatch(s);
    if (paren != null) {
      year = int.tryParse(RegExp(r'(19|20)\d{2}').firstMatch(paren.group(0)!)!.group(0)!);
      s = '${s.substring(0, paren.start)} ${s.substring(paren.end)}'.trim();
    } else {
      final tail = RegExp(r'[\s\-\(\[]?(19|20)\d{2}[\)\]]?\s*$').firstMatch(s);
      if (tail != null) {
        year = int.tryParse(RegExp(r'(19|20)\d{2}').firstMatch(tail.group(0)!)!.group(0)!);
        s = s.substring(0, tail.start).trim();
      }
    }
    // Drop common quality / version tokens wherever they sit.
    final qualityTokens = RegExp(
      r'\b(FHD|HD|SD|4K|UHD|HEVC|HDR|H\.?264|H\.?265|X\.?264|X\.?265|'
      r'BluRay|BRRip|WEB[-.]?DL|WEBRip|HDTV|DVDRip|REMUX|MULTI|DV|'
      r'VOSTFR|VOST|STFR|VFQ|VFF|VF|VO)\b',
      caseSensitive: false,
    );
    s = s.replaceAll(qualityTokens, '');
    // Bracket / pipe separators → space.
    s = s.replaceAll(RegExp(r'[\(\)\[\]\|]+'), ' ');
    // Per-token cleanup: strip edge dashes (drops standalone "-" and the
    // trailing dash left by "Title - 2025") while preserving intra-word
    // dashes like "9-1-1".
    s = s
        .split(RegExp(r'\s+'))
        .map((t) => t.replaceAll(RegExp(r'^[-–]+|[-–]+$'), ''))
        .where((t) => t.isNotEmpty)
        .join(' ')
        .trim();

    return TitleYear(title: s, year: year, original: raw);
  }
}

/// Result of [TitleYearParser.parse].
class TitleYear {
  final String title;
  final int? year;
  final String original;

  const TitleYear({
    required this.title,
    required this.year,
    required this.original,
  });

  bool get isUsable => title.isNotEmpty;

  @override
  String toString() => 'TitleYear($title, $year)';
}

File: tvos/UniStreamTV/UniStreamTV/Services/TitleYearParser.swift
import Foundation

/// Parses raw IPTV titles into a clean `(title, year)` pair for TMDB lookups.
///
/// Handles common IPTV prefixes and quality tokens:
///   "FR| Fire Country - 2025"       -> ("Fire Country", 2025)
///   "FR| Fire Country (2025)"       -> ("Fire Country", 2025)
///   "SD| Cold Wallet FHD - 2023"    -> ("Cold Wallet", 2023)
///   "FHD Fire Country"              -> ("Fire Country", nil)
///   "=FR= FRANCE =FR="              -> ("FRANCE", nil)  (best-effort)
///
/// Mirror of the Dart `TitleYearParser` in the Flutter codebase so the two
/// platforms yield the same cache keys when they share a server.
enum TitleYearParser {

    struct Result {
        let title: String
        let year: Int?
        let original: String
        var isUsable: Bool { !title.isEmpty }
    }

    static func parse(_ raw: String) -> Result {
        let original = raw
        var s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.isEmpty { return Result(title: "", year: nil, original: original) }

        // Drop leading pipe-wrapped tag blocks: "|FR|", "|FR-4K DV|", "|IT-4K|",
        // and stacked ones like "|VO|STFR|". Some panels wrap the country/quality
        // tag in pipes (leading pipe first), e.g. "|FR| Extravagances (1995)".
        // Without this the "FR" leaks into the title and breaks TMDB matching.
        while let r = s.range(of: #"^\s*\|[^|]*\|\s*"#, options: .regularExpression) {
            s.removeSubrange(r)
        }

        // Drop a remaining country / quality prefix without a leading pipe:
        // "FR|", "SD|", "STFR|", "VO|", "BE|", "UK|", "US|" …
        if let match = s.range(
            of: #"^[A-Z]{2,4}[-\s]?\|\s*"#,
            options: [.regularExpression, .caseInsensitive]
        ) {
            s.removeSubrange(match)
        }

        // Cut trailing junk after a "||" separator, e.g.
        // "La casa de papel (2017) || MULTI".
        if let dbl = s.range(of: "||") {
            s = String(s[..<dbl.lowerBound]).trimmingCharacters(in: .whitespacesAndNewlines)
        }

        // Extract the year. Prefer a parenthesised/bracketed 4-digit year anywhere
        // ("(2017)", "[2020]") — robust against trailing tags like "(VFQ)" that
        // follow the year. Fall back to a bare trailing year (" - 2025", " 2025").
        var year: Int?
        if let paren = s.range(of: #"[\(\[](19|20)\d{2}[\)\]]"#, options: .regularExpression) {
            let slice = s[paren]
            if let digits = slice.range(of: #"(19|20)\d{2}"#, options: .regularExpression) {
                year = Int(slice[digits])
            }
            s = (String(s[..<paren.lowerBound]) + " " + String(s[paren.upperBound...]))
                .trimmingCharacters(in: .whitespacesAndNewlines)
        } else if let tail = s.range(
            of: #"[\s\-\(\[]?(19|20)\d{2}[\)\]]?\s*$"#,
            options: .regularExpression
        ) {
            let slice = s[tail]
            if let digits = slice.range(of: #"(19|20)\d{2}"#, options: .regularExpression) {
                year = Int(slice[digits])
            }
            s.removeSubrange(tail)
            s = s.trimmingCharacters(in: .whitespacesAndNewlines)
        }

        // Drop common quality / version tokens wherever they sit.
        let qualityPattern =
            #"\b(FHD|HD|SD|4K|UHD|HEVC|HDR|H\.?264|H\.?265|X\.?264|X\.?265|"# +
            #"BluRay|BRRip|WEB[-.]?DL|WEBRip|HDTV|DVDRip|REMUX|MULTI|DV|"# +
            #"VOSTFR|VOST|STFR|VFQ|VFF|VF|VO)\b"#
        s = s.replacingOccurrences(
            of: qualityPattern,
            with: "",
            options: [.regularExpression, .caseInsensitive]
        )

        // Bracket / pipe separators → space (intra-word dashes like "9-1-1" kept).
        s = s.replacingOccurrences(of: #"[\(\)\[\]\|]+"#, with: " ", options: .regularExpression)
        // Per-token cleanup: strip edge dashes (drops standalone "-" and the
        // trailing dash left by "Title - 2025") while preserving intra-word dashes.
        s = s.split(separator: " ", omittingEmptySubsequences: true)
            .map { token in
                token.replacingOccurrences(
                    of: #"^[-–]+|[-–]+$"#, with: "", options: .regularExpression
                )
            }
            .filter { !$0.isEmpty }
            .joined(separator: " ")
        s = s.trimmingCharacters(in: .whitespacesAndNewlines)

        return Result(title: s, year: year, original: original)
    }
}
